### DISCUSSION TITLE
How to implement recursion visulaizer

### DISCUSSION THREAD

**Discussion by istaari** [community/community#127666] [Other Feature Feedback, Questions, & Ideas] [answered]
I want to create a web app that visualizes recursion calls in vanilla JS. For instance, when a user inputs a recursive function, the app generates a recursion tree of each recursive call.

```javascript
function factorial(n) {
  if (n === 0 || n === 1) {
    return 1;
  } else {
    return n * factorial(n - 1);
  }
}
```
The app should generate a recursion call tree for each step. Do you have any ideas or suggestions for how I should approach this?

</br>
</br>

 **_Guidelines_**

 I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by vishal11u** [comment]
function factorial(n, depth = 0) {
  // Log the call with appropriate indentation
  console.log(`${' '.repeat(depth * 2)}factorial(${n})`);

  // Base case
  if (n === 0 || n === 1) {
    console.log(`${' '.repeat(depth * 2)}return 1`);
    return 1;
  }

  // Recursive case
  const result = n * factorial(n - 1, depth + 1);

  // Log the result with appropriate indentation
  console.log(`${' '.repeat(depth * 2)}return ${result}`);
  return result;
}

// Example usage
factorial(5);

## output
factorial(5)
  factorial(4)
    factorial(3)
      factorial(2)
        factorial(1)
        return 1
      return 2
    return 6
  return 24
return 120

**Comment by istaari** [comment]
@vishal11u thanks for the input.

**Reply by vishal11u** [reply]
so give my answer as correct answer plz

**Reply by istaari** [reply, marked as answer]
Hey, I was looking for something like this. Anyway, I found it.

**Recursion Tree Visualizer**

App Link : (https://recursion.vercel.app/)
Github Link : https://github.com/brpapa/recursion-tree-visualizer
